linera-rpc/src/simple/server.rs

```diff
@@ -83,62 +85,38 @@ where
         cross_chain_sender_delay: Duration,
         cross_chain_sender_failure_rate: f32,
         this_shard: ShardId,
-        mut receiver: mpsc::Receiver<(RpcMessage, ShardId)>,
+        receiver: mpsc::Receiver<(CrossChainRequest, ShardId)>,
     ) {
-        let mut pool = network
-            .protocol
-            .make_outgoing_connection_pool()
-            .await
-            .expect("Initialization should not fail");
-
-        'receive_loop: while let Some((message, shard_id)) = receiver.next().await {
-            if cross_chain_sender_failure_rate > 0.0
-                && rand::thread_rng().gen::<f32>() < cross_chain_sender_failure_rate
-            {
-                warn!("Dropped 1 cross-message intentionally.");
-                continue;
-            }
-
-            let shard = network.shard(shard_id);
-            let remote_address = format!("{}:{}", shard.host, shard.port);
-
-            // Send the cross-chain query and retry if needed.
-            for i in 0..cross_chain_max_retries {
-                // Delay increases linearly with the attempt number.
-                linera_base::time::timer::sleep(
-                    cross_chain_sender_delay + cross_chain_retry_delay * i,
-                )
-                .await;
-
-                let status = pool.send_message_to(message.clone(), &remote_address).await;
-                match status {
-                    Err(error) => {
-                        warn!(
-                            nickname,
-                            %error,
-                            i,
-                            from_shard = this_shard,
-                            to_shard = shard_id,
-                            "Failed to send cross-chain query",
-                        );
-                    }
-                    _ => {
-                        debug!(
-                            from_shard = this_shard,
-                            to_shard = shard_id,
-                            "Sent cross-chain query",
-                        );
-                        continue 'receive_loop;
-                    }
+        let pool = Arc::new(Mutex::new(
```

Comment:
Alternatively we could make the pool clonable:
* In the UDP case that would simply mean putting the mutex _inside_ the pool struct.
* In the TCP case we could use a `DashMap`, just like we do for gRPC.

Reply:
I'm not fussed either way.  In both cases the changes are limited to `linera_rpc::simple` and have no wider impact.  Probably as a general rule I prefer to leave the mutex outside an opaque struct so that consumers can choose to opt in/out.

Reply:
It's just that with the `DashMap`, multiple requests could be handled at the same time. Keeping the mutex outside makes that impossible.

Reply:
Seems complicated, though; I think we'd even need mutexes in the `DashMap`'s entries. I'll leave that for another PR, if we decide to do it.